typing/gprinttyp.ml

```diff
@@ -0,0 +1,872 @@
+
+  let merge_label l r =
+    let r' = String_set.of_list r in
+    let l' = String_set.of_list l in
+    List.filter (fun x -> not (String_set.mem x r') ) l
+    @ List.filter (fun x -> not (String_set.mem x l') ) r
+
+  let merge l r =
+    { color = alt l.color r.color;
+      style = alt l.style r.style;
+      label = merge_label l.label r.label;
+      font_color = alt l.font_color r.font_color;
+      shape = alt l.shape r.shape;
+    }
+  let txt t = Label [t]
+
+end
+type decoration = Decoration.r
+
+type dir = Toward | From
+
+let txt = Decoration.txt
+let std = Decoration.none
+let dotted = Decoration.(make [Style Dotted])
+let memo = Decoration.(make [txt "expand"; Style Dash] )
+
+
+type params = {
+  short_ids:bool;
+  ellide_links:bool;
+  expansion_as_hyperedge:bool;
+  colorize:bool;
+}
+
+let ellide_links ty =
+  let rec follow_safe visited t =
+    let t = Types.Transient_expr.coerce t in
+    if List.memq t visited then t
+    else match t.Types.desc with
+      | Tlink t' -> follow_safe (t::visited) t'
+      | _ -> t
+  in
+  follow_safe [] ty
+
+let repr params ty =
+  if params.ellide_links then ellide_links ty
+  else Types.Transient_expr.coerce ty
+
+let colorize params id =
+  if not params.colorize then None
+  else
+    let nhues = 200 in
+    let h = float_of_int (17 * id mod nhues) /. float_of_int nhues in
+    let s = match id mod 3 with
+      | 0 -> 0.3
+      | 1 -> 0.5
+      | 2 | _ -> 0.7
+     in
+     let l = match id mod 7 with
+       | 0 -> 0.5
+       | 1 -> 0.55
+       | 2 -> 0.60
+       | 3 -> 0.65
+       | 4 -> 0.70
+       | 5 -> 0.75
+       | 6 | _ -> 0.8
+     in
+     Some (Decoration.hsl ~h ~s ~l)
+
+module Index: sig
+  type t = private
+    | Main of int
+    | Synthetic of int
+    | Field of { id:int; synth:bool; name:string }
+  val color_id: t -> int
+  val field: name:string -> t -> t
+  val field_ext: Types.row_field_cell ->  t
+  val split:
+    params -> Types.type_expr -> t * Decoration.color option * Types.type_desc
+ end = struct
+
+  type t =
+    | Main of int
+    | Synthetic of int
+    | Field of { id:int; synth:bool; name:string }
+
+  type name_map = {
+    last: int ref;
+    tbl: (int,int) Hashtbl.t;
+  }
+  let id_map = { last = ref 0; tbl = Hashtbl.create 20 }
+
+  let color_id = function
+    | Main id | Synthetic id | Field {id;_} -> id
+
+
+  let pretty_id params id =
+    if not params.short_ids then Main id else
+      match Hashtbl.find_opt id_map.tbl id with
+      | Some x -> Main x
+      | None ->
+          incr id_map.last;
+          let last = !(id_map.last) in
+          Hashtbl.replace id_map.tbl id last;
+          Main last
+
+  let split params x =
+    let x = repr params x in
+    let color = colorize params x.id in
+    pretty_id params x.id, color, x.desc
+
+  let _index params ty = pretty_id params (repr params ty).id
+  let field ~name x = match x with
+    | Main id -> Field {id;name;synth=false}
+    | Field r -> Field {r with name}
```

Comment:
I don't understand the logic here -- I'm not sure what `field` is and from a distance I would have expected that the field `baz` of `foo.bar` is `foo.bar.baz`, not `foo.baz`. No big deal, but maybe a comment. Also when reading that I wondered if we should have `Field of t * string` instead of your weird definition, but I'm fine with the weird definition as well.